libafl_frida/src/alloc.rs

```diff
@@ -0,0 +1,408 @@
+                    page_size,
+                    ProtFlags::PROT_READ | ProtFlags::PROT_WRITE,
+                    MapFlags::MAP_PRIVATE
+                        | MapFlags::MAP_ANONYMOUS
+                        | MapFlags::MAP_FIXED
+                        | MapFlags::MAP_NORESERVE,
+                    -1,
+                    0,
+                )
+            }
+            .is_ok()
+            {
+                shadow_bit = *try_shadow_bit;
+                break;
+            }
+        }
+        assert!(shadow_bit != 0);
+
+        // attempt to pre-map the entire shadow-memory space
+        let addr: usize = 1 << shadow_bit;
+        let pre_allocated_shadow = unsafe {
+            mmap(
+                addr as *mut c_void,
+                addr + addr,
+                ProtFlags::PROT_READ | ProtFlags::PROT_WRITE,
+                MapFlags::MAP_ANONYMOUS
+                    | MapFlags::MAP_FIXED
+                    | MapFlags::MAP_PRIVATE
+                    | MapFlags::MAP_NORESERVE,
+                -1,
+                0,
+            )
+        }
+        .is_ok();
+
+        Self {
+            options,
+            page_size,
+            pre_allocated_shadow,
+            shadow_offset: 1 << shadow_bit,
+            shadow_bit,
+            allocations: HashMap::new(),
+            shadow_pages: RangeSet::new(),
+            allocation_queue: HashMap::new(),
+            largest_allocation: 0,
+            base_mapping_addr: addr + addr + addr,
+            current_mapping_addr: addr + addr + addr,
+        }
+    }
+
+    /// Retreive the shadow bit used by this allocator.
+    pub fn shadow_bit(&self) -> u32 {
+        self.shadow_bit as u32
+    }
+
+    #[inline]
+    fn round_up_to_page(&self, size: usize) -> usize {
+        ((size + self.page_size) / self.page_size) * self.page_size
+    }
+
+    #[inline]
+    fn round_down_to_page(&self, value: usize) -> usize {
+        (value / self.page_size) * self.page_size
+    }
+
+    fn find_smallest_fit(&mut self, size: usize) -> Option<AllocationMetadata> {
+        let mut current_size = size;
+        while current_size <= self.largest_allocation {
+            if self.allocation_queue.contains_key(&current_size) {
+                if let Some(metadata) = self.allocation_queue.entry(current_size).or_default().pop()
+                {
+                    return Some(metadata);
+                }
+            }
+            current_size *= 2;
+        }
+        None
+    }
+
+    pub unsafe fn alloc(&mut self, size: usize, _alignment: usize) -> *mut c_void {
+        let mut is_malloc_zero = false;
+        let size = if size == 0 {
+            println!("zero-sized allocation!");
+            is_malloc_zero = true;
+            16
+        } else {
+            size
+        };
+        if size > (1 << 30) {
+            panic!("Allocation is too large: 0x{:x}", size);
+        }
+        let rounded_up_size = self.round_up_to_page(size) + 2 * self.page_size;
```

Comment:
Is it fine if the size is now `1 << 30`?

Reply:
This was an arbitrary selection. It should really be configurable... I'll open an issue...